src/common/task.cc

```diff
@@ -72,10 +72,10 @@ class TaskBuilder {
   }
 
   void SetRequiredResource(int64_t resource_index, double value) {
-    if (resource_index >= resource_vector_.size()) {
+    if (static_cast<size_t>(resource_index) >= resource_vector_.size()) {
       /* Make sure the resource vector is constructed entry by entry,
        * in order. */
-      CHECK(resource_index == resource_vector_.size());
+      CHECK(static_cast<size_t>(resource_index) == resource_vector_.size());
       resource_vector_.resize(resource_index + 1);
```

Comment:
This is especially confusing... are you adding 1 to a potentially negative value? Is that what you intend?

Reply:
Are you saying that in general casting from signed to unsigned is bad? E.g., would it be better to do the following?

```c++
CHECK(resource_index == static_cast<int64_t>(resource_vector_.size()));
```

Reply:
No, I'm saying (like in the [previous comment](//github.com/ray-project/ray/pull/1116#discussion_r144410877)) the variable should be unsigned. Right now you're adding 1 to a signed value and then resizing based on that, which doesn't make sense in the negative case.

Reply:
So in our case we have decided to make all index variables unsigned and we try to keep it that way; this is also done by others like Google and Arrow so we are in good company. The reason is that it simplifies arithmetic and guards against the variable wrapping around zero (it is easier to detect a negative index than a large positive one).

Reply:
Signed values don't actually simplify anything. (?) They only make it complicated to range-check, since now you have to check both `< 0` and `>= size()` rather than just the latter. I'm also not sure
 hwat you mean about "guarding" against variables wrapping around zero—adding `(ptrdiff_t)(-1)` and adding `~(size_t)0` to a pointer `p` results in exactly the same address `p - 1` regardless of which one you choose. In fact, if you have checked pointers enabled (`_GLIBCXX_DEBUG` may do this? not sure) they may actually catch large unsigned values since those would always be illegal even if they were to land inside the array, but they cannot possibly catch the bitwise-identical negative values if they happen to point inside the array.

Moreover, what there _is_ a difference in is that overflowing signed values is in fact _undefined_ rather than _well-defined_ behavior. So that means your code could do all sorts of crazy things if you in fact do happen to accidentally overflow, rather than merely writing to a lower spot than the one you intended to.

More generally, I would not hold Google as the example of ideal C++ coding. They have guidelines that are pretty universally understood to be often bad advice (no need to take my word for this; just Google around...), and tailored to their own particular use cases. I don't know why Arrow might do this but I'm sure you won't have trouble finding people who use unsigned values either... =P

Reply:
That said, you obviously don't have to change this if you'd rather not. Just letting you know what the issues are.